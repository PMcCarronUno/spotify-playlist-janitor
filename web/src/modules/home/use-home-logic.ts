import { useCallback, useEffect, useState } from "react";

import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";

import { Playlist } from "./home-types";
import { useModal } from "shared/components/modal";

export const useHomeLogic = () => {
  const { getDatabasePlaylists, addDatabasePlaylist } = useDataApi();
  const { getSpotifyPlaylists } = useSpotifyApi();

  const [loading, setLoading] = useState(false);
  const [modalSaving, setModalSaving] = useState(false);
  const [showSpotifyAuthModal, setShowSpotifyAuthModal] = useState(false);
  const [monitoredPlaylists, setMonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [unmonitoredPlaylists, setUnmonitoredPlaylists] = useState<
    Playlist[] | undefined
  >(undefined);
  const [modalError, setModalError] = useState<string | undefined>(undefined);

  const { isOpen: modalOpen, onOpen: onModalOpen, onClose } = useModal();

  const onModalClose = useCallback(() => {
    setModalError(undefined);
    onClose();
  }, [setModalError, onClose]);

  const getPlaylistData = useCallback(async () => {
    try {
      setLoading(true);
      const databasePlaylistsResponse = await getDatabasePlaylists();
      const spotifyPlaylistsResponse = await getSpotifyPlaylists();

      setMonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter((spotifyPlaylist) =>
            databasePlaylistsResponse.data.some(
              (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
            )
          )
          .map((monitoredPlaylist) => ({
            id: monitoredPlaylist.id,
            href: monitoredPlaylist.href,
            name: monitoredPlaylist.name,
            image: monitoredPlaylist.images[0],
          }))
      );

      setUnmonitoredPlaylists(
        spotifyPlaylistsResponse.data
          .filter(
            (spotifyPlaylist) =>
              !databasePlaylistsResponse.data.some(
                (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
              )
          )
          .map((unmonitoredPlaylist) => ({
            id: unmonitoredPlaylist.id,
            href: unmonitoredPlaylist.href,
            name: unmonitoredPlaylist.name,
            image: unmonitoredPlaylist.images[0],
          }))
      );
      setLoading(false);
    } catch (e: any) {
      if (
        e?.response?.status === 500 &&
        e?.response?.data?.message ===
          "Application has not been logged into your Spotify account."
      ) {
        setShowSpotifyAuthModal(true);
      }
      setLoading(false);
    }
  }, [
    setLoading,
    getDatabasePlaylists,
    getSpotifyPlaylists,
    setMonitoredPlaylists,
    setUnmonitoredPlaylists,
  ]);

  useEffect(() => {
    getPlaylistData();
  }, [getPlaylistData]);

  const onSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      setModalError(undefined);
      const playlistId = (event.currentTarget[1] as HTMLInputElement).value;
      const skipThreshold = (event.currentTarget[2] as HTMLInputElement).value;
      const ignoreInitialSkips = (event.currentTarget[3] as HTMLInputElement)
        .checked;
      const autoDeleteTracksAfter = (event.currentTarget[4] as HTMLInputElement)
        .value;

      if (!playlistId) {
        setModalError("Please select a playlist");
        return;
      }
      if (!!autoDeleteTracksAfter && parseInt(autoDeleteTracksAfter) === 0) {
        setModalError("Auto-delete must be empty or greater than 0");
        return;
      }

      try {
        setModalSaving(true);
        
        const request = {
          id: playlistId,
          skipThreshold: !!skipThreshold ? parseInt(skipThreshold) : undefined,
          ignoreInitialSkips: ignoreInitialSkips,
          autoCleanupLimit: !!autoDeleteTracksAfter
            ? parseInt(autoDeleteTracksAfter)
            : undefined,
        };

        await addDatabasePlaylist(request);
        setModalSaving(false);
        onModalClose();
        await getPlaylistData();
      } catch (e: any) {
        setModalError(e.response?.data?.message || "Unknown error");
        setModalSaving(false);
      }
    },
    [addDatabasePlaylist, getPlaylistData, onModalClose]
  );

  const onPlaylistChange = useCallback(
    async (newValue: { label: string; value: string }) => {
      if (newValue?.value) {
        setModalError(undefined);
      }
    },
    []
  );

  return {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } as const;
};
