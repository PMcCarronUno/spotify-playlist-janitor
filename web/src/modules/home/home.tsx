import styled from "styled-components";
import { AiOutlinePlusCircle } from "react-icons/ai";

import { BLACK, GREEN, RED } from "shared/constants";
import { SubTitle, Text } from "shared/components/typography";
import { Modal } from "shared/components/modal";

import { useHomeLogic } from "./use-home-logic";
import { AddPlaylistModalView } from "./modal/add-playlist-modal";

export const Home = () => {
  const {
    monitoredPlaylists,
    unmonitoredPlaylists,
    loading,
    modalOpen,
    onModalOpen,
    onModalClose,
    onSubmit,
    onPlaylistChange,
    modalError,
    modalSaving,
    showSpotifyAuthModal,
  } = useHomeLogic();

  return (
    <PageContainer>
      {monitoredPlaylists && (
        <>
          {monitoredPlaylists.map((monitoredPlaylist, index) => (
            <div
              key={monitoredPlaylist.id}
              className="item"
              onClick={() =>
                alert(
                  `Naviagte to ${monitoredPlaylist.name} at route ${monitoredPlaylist.id}`
                )
              }
              id={`playlist-item-${index}`}
              data-testid={`playlist-item-${index}`}
            >
              {monitoredPlaylist.image && (
                <img
                  alt={monitoredPlaylist.name}
                  width={130}
                  height={130}
                  src={monitoredPlaylist.image.url}
                />
              )}

              <SubTitle>{monitoredPlaylist.name}</SubTitle>
            </div>
          ))}
          <div
            className="item new"
            onClick={onModalOpen}
            id={`add-playlist-item`}
            data-testid={`add-playlist-item`}
          >
            <AiOutlinePlusCircle />
          </div>
        </>
      )}
      <Modal
        {...{
          isOpen: modalOpen,
          onClose: onModalClose,
          label: "Select Playlist Modal",
        }}
      >
        <AddPlaylistModalView
          {...{
            onSubmit,
            unmonitoredPlaylists,
            onPlaylistChange,
            modalSaving,
            modalError,
            onModalClose,
          }}
        />
      </Modal>
      <Modal
        {...{
          isOpen: showSpotifyAuthModal,
          onClose: () => {},
          label: "Spotify Auth Modal",
        }}
      >
        <ModalContainer>
          <SubTitle style={{ marginBottom: 8 }}>
            Spotify Authentication Error
          </SubTitle>
          <Text>
            The application has not been authenticated with your Spotify
            account. Please follow{" "}
            <a
              href={process.env.REACT_APP_API_URL}
              target="_blank"
              rel="noreferrer"
            >
              this
            </a>{" "}
            link and sign into Spotify, and <a href=".">reload</a> this page.
          </Text>
        </ModalContainer>
      </Modal>
    </PageContainer>
  );
};

const PageContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  align-content: flex-start;

  color: white;
  width: 100%;
  max-width: 1825px;
  paddingtop: 18px;

  .item {
    flex: 1 1 30%; /*grow | shrink | basis */
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    margin: 8px;
    height: 160px;
    max-width: 33%;
    background-color: ${BLACK};

    border-radius: 15px;

    cursor: pointer;

    &:hover {
      transform: scale(1.01);
      box-shadow: 0 0 32px rgba(117, 117, 117, 0.2);
    }
  }

  .new {
    background-color: ${GREEN};

    svg {
      height: 6em;
      width: 6em;
    }
  }
`;

const ModalContainer = styled.div`
  width: 360px;

  .rows {
    margin-bottom: 8px;
  }

  .row {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-bottom: 8px;
  }

  .number-input {
    width: 55px !important;
  }

  .toggle-input {
    margin-right: 16px;
  }

  .auto-delete-input {
    margin-right: -10px;
  }

  .auto-delete-post-text {
    margin-right: 10px;
  }

  .error {
    color: ${RED};
  }
`;

Home.displayName = "Home";
