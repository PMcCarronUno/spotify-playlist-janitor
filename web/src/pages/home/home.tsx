import styled from "styled-components";
import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";
import { Home as HomeInitialise } from "modules/home";

export const Home = () => {
  return (
    <Content>
      <HomeInitialise />
    </Content>
  );
};

const Content = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
`;

Home.displayName = "Home";
