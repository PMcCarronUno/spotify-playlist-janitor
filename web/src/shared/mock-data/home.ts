import { Playlist } from "modules/home/home-types";
import { databasePlaylists, spotifyPlaylists } from "shared/mock-data/api";

export const monitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter((spotifyPlaylist) =>
    databasePlaylists.some(
      (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
    )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));

export const unmonitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter(
    (spotifyPlaylist) =>
      !databasePlaylists.some(
        (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
      )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));
