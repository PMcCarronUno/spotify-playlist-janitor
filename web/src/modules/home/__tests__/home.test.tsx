import {
  fireEvent,
  getByTestId,
  render,
  waitFor,
} from "@testing-library/react";

import { Home } from "../home";
import { useHomeLogic } from "../use-home-logic";
import {
  monitoredPlaylists,
  unmonitoredPlaylists,
} from "shared/mock-data/home";

jest.mock("../use-home-logic");
jest.mock("../modal");

describe("<Home />", () => {
  let container: HTMLElement;

  const mockUseHomeLogic = {
    monitoredPlaylists: monitoredPlaylists,
    unmonitoredPlaylists: unmonitoredPlaylists,
    loading: false,
    modalOpen: false,
    onModalOpen: jest.fn(),
    onModalClose: jest.fn(),
    onSubmit: jest.fn(),
    onPlaylistChange: jest.fn(),
    modalError: "",
    modalSaving: false,
    showSpotifyAuthModal: false,
  };

  beforeEach(async () => {
    jest.mocked(useHomeLogic).mockImplementation(() => mockUseHomeLogic);

    await waitFor(() => {
      ({ container } = render(<Home />));
    });
  });

  it("should render Home component", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render Home component with no monitored playlists", async () => {
    const mockUseHomeLogic2 = {
      ...mockUseHomeLogic,
      monitoredPlaylists: [],
    };
    jest.mocked(useHomeLogic).mockImplementation(() => mockUseHomeLogic2);

    await waitFor(() => {
      ({ container } = render(<Home />));
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onModalOpen function when Add button clicked", () => {
    const addPlaylist = getByTestId(container, "add-playlist-item");
    fireEvent.click(addPlaylist);

    expect(mockUseHomeLogic.onModalOpen).toHaveBeenCalled();
  });
});
