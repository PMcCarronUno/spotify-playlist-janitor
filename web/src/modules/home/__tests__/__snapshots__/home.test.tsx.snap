// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render Home component 1`] = `
<div
  class="sc-eTNxZ cjXmkm"
>
  <div
    class="item playlist"
    data-testid="playlist-item-0"
    id="playlist-item-0"
  >
    <img
      alt="Playlist 1"
      height="130"
      src="https://mosaic.scdn.co/101"
      width="130"
    />
    <div
      class="sc-guDMob jLqGJD"
    >
      Playlist 1
    </div>
  </div>
  <div
    class="item playlist"
    data-testid="playlist-item-1"
    id="playlist-item-1"
  >
    <img
      alt="Playlist 2"
      height="130"
      src="https://mosaic.scdn.co/201"
      width="130"
    />
    <div
      class="sc-guDMob jLqGJD"
    >
      Playlist 2
    </div>
  </div>
  <div
    class="item playlist"
    data-testid="playlist-item-2"
    id="playlist-item-2"
  >
    <img
      alt="Playlist 3"
      height="130"
      src="https://mosaic.scdn.co/301"
      width="130"
    />
    <div
      class="sc-guDMob jLqGJD"
    >
      Playlist 3
    </div>
  </div>
  <div
    class="item playlist"
    data-testid="playlist-item-3"
    id="playlist-item-3"
  >
    <img
      alt="Playlist 4"
      height="130"
      src="https://mosaic.scdn.co/401"
      width="130"
    />
    <div
      class="sc-guDMob jLqGJD"
    >
      Playlist 4
    </div>
  </div>
  <div
    class="item playlist new "
    data-testid="add-playlist-item"
    id="add-playlist-item"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
      />
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
      />
    </svg>
  </div>
</div>
`;

exports[`<Home /> should render Home component with empty page 1`] = `
<div
  class="sc-eTNxZ cjXmkm"
>
  <div
    data-testid="empty-loading"
  />
</div>
`;

exports[`<Home /> should render Home component with loading skeletons 1`] = `
<div
  class="sc-eTNxZ cjXmkm"
>
  <div
    class="item"
    data-testid="playlist-item-skeleton-0"
    id="playlist-item-skeleton-0"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 100%; border-radius: 15px; --base-color: #020202; --highlight-color: #444;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
  <div
    class="item"
    data-testid="playlist-item-skeleton-1"
    id="playlist-item-skeleton-1"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 100%; border-radius: 15px; --base-color: #020202; --highlight-color: #444;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
  <div
    class="item"
    data-testid="playlist-item-skeleton-2"
    id="playlist-item-skeleton-2"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 100%; border-radius: 15px; --base-color: #020202; --highlight-color: #444;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
  <div
    class="item"
    data-testid="playlist-item-skeleton-3"
    id="playlist-item-skeleton-3"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 100%; border-radius: 15px; --base-color: #020202; --highlight-color: #444;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
  <div
    class="item"
    data-testid="playlist-item-skeleton-4"
    id="playlist-item-skeleton-4"
  >
    <span
      aria-busy="true"
      aria-live="polite"
    >
      <span
        class="react-loading-skeleton"
        style="height: 100%; border-radius: 15px; --base-color: #020202; --highlight-color: #444;"
      >
        ‌
      </span>
      <br />
    </span>
  </div>
  <div
    class="item playlist new new-disabled"
    data-testid="add-playlist-item"
    id="add-playlist-item"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
      />
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
      />
    </svg>
  </div>
</div>
`;

exports[`<Home /> should render Home component with no monitored playlists 1`] = `
<div
  class="sc-eTNxZ cjXmkm"
>
  <div
    class="item playlist new "
    data-testid="add-playlist-item"
    id="add-playlist-item"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 1024 1024"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M696 480H544V328c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v152H328c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h152v152c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V544h152c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8z"
      />
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
      />
    </svg>
  </div>
</div>
`;
